slac-lcls/lute
# ENH Configuration Changes for Usage of Templates and Database Access

# Description

This PR introduces:
- Reorganization of pydantic parameter models
  - All `Task`s will share an instance of `AnalysisHeader` which contains general lute configuration parameters.
  - Third party task models now inherit from `BaseBinaryParameters`
  - `ThirdPartyParameter` objects are used to represent items which should be used for third party configuration files.
  - `TemplateConfig` models are used for configuring how to compile templates for third party configuration files.
  - Changes in validators to configure some of the above models.
  - Models dependent on other task parameters can use the database API to retrieve values.
- Changes in the `BinaryTask` class
  - Will manage parameters passed through the appropriate pydantic model objects to prepare a third party configuration file using jinja.
- Changes in the `BaseExecutor` and `Executor` classes to store and retrieve `DescribedAnalysis` objects on a task-by-task basis in the database.
- A new database is used for storing parameter information on a task-by-task, submission-by-submission basis. A simplified, backend-agnostic, API is provided for basic storage and retrieval tasks. The current version uses `sqlite` for the database itself.
- Task schemas
  - `Task`s declare themselves to implement/conform to specific schemas which categorize the type of task they are and what kind of output can be expected.

## Checklist
- [x] Changes for `lute.io.config`
  - [x] `AnalysisHeader` 
  - [x] `ThirdPartyParameters`
  - [x] `TemplateConfig`
  - [x] `BaseBinaryParameters` 
  - [x] Change affected third party tasks and update validators 
  - [x] Update `parse_config` to implement the new model changes.
- [x] `sqlite` database for parameter storage
  - [x] Backend database design
  - [x] Standardized read/write operations
  - [x] High-level API
- [x] `BaseExecutor` and `Executor` changes for database accesses 
- [x] Changes to example configuration YAMLs
- [x] Miscellaneous signaling changes for handling posix signals.
- [x] Jinja template for `smalldata_tools`' `producer` file. 
- [x] `Task` schemas placeholder
- [x] Refactor and reorganize to avoid circular dependencies
- [x] Improve IPC cleanup with context managers and cleanup functions
  - Avoid relying on garbage collection (currently not using weak references) 
- [x] Read and Write test tasks which demonstrate retrieving data from the database. 
- [x] Rename `AnalysisConfig` to `DescribedAnalysis` 
- [x] Provide `dbview` utility - A readonly TUI built with `textual` to view LUTE database entries. 
  - Usage: `./dbview -p path/to/lute.db`

## PR Type:

## Address issues:
- NA

# Testing
- A database will be created for any Task when using this branch. The database will be written to `lute.db` in the `work_dir` directory defined in the configuration YAML (first document - the header).
- Two specific task were also written to test whether a file could be written out, the location recorded in the database and then the file retrieved for a subsequent task.
- In order to run these tests:
  1. `cd` to `/sdf/home/d/dorlhiac/scratch/lute_tests/lute_dorlhiac`
  2. Checkout this branch (`ENH/templates_and_db`)
  3. Edit line 8 of `config/test.yaml` (`work_dir: "/sdf/scratch/..."`) to point to where you want the database and output files to be. `/sdf/home/d/dorlhiac/scratch/lute_tests/lute_dorlhiac` is fine.
  4. `cd ..`
  5. In `/sdf/home/d/dorlhiac/scratch/lute_tests` run: `sbatch test_writeoutput.sh` (Run *first* to write a result.)
  6. Then run `sbatch test_readoutput.sh`

Your output for the write task will be in `write_test.err`, and the output for the read task in `readoutput_test.err`. There will also be a `lute.db` database, and a `test_output.txt` file in the directory as well. `readoutput_test.err`'s log file should include the input file in the configuration:
```
INFO:lute.execution.executor:lute_config=AnalysisHeader(title='LUTE Task Configuration', experiment='EXPL10000', run='', date='2023/10/25', lute_version=0.1, task_timeout=600, work_dir='/sdf/scratch/users/d/dorlhiac') in_file='/sdf/scratch/users/d/dorlhiac/test_output.txt'
INFO:lute.execution.executor:Successfully loaded data!
INFO:lute.execution.executor:Was able to load data.
```

If the steps above were followed, `in_file` should not have been provided by anything other than the database. For both the read and write Tasks you will also see some debug lines with SQL queries.


To run most tests in order try:
```bash
sbatch test_first_party.sh
sbatch test_binary.sh
sbatch test_socket.sh
sbatch test_writeoutput.sh
sbatch test_readoutput.sh
```

Smalldata can also be run; however, with the new templating system be careful about how you are running it and where you are outputing the files!
# Screenshots
NA
# Screenshots
NA
